static/js/jira-manual-app-creation.js

```diff
@@ -1,79 +1,150 @@
+		if (child.closed) {
+			clearInterval(interval);
+			AP.navigator.go(
+				'addonmodule',
+				{
+					moduleKey: "github-post-install-page"
+				}
+			);
+		}
+	}, 1000);
+
+	return child;
+}
+
+const handleFormErrors = (isUpdate) => {
+	$(".jiraManualAppCreation__serverError").show();
+	$(".errorMessageBox__message").empty().append("Please make sure all the details you entered are correct.");
+	const errorTitle = ".errorMessageBox__title";
+
+	isUpdate
+		? $(errorTitle).empty().append("We couldn't update your GitHub app.")
+		: $(errorTitle).empty().append("We couldn't create your GitHub app.");
+}
+
+const gitHubAppPutRequest = (uuid, isUpdate, data) => {
+	$.ajax({
+		type: "PUT",
+		url: `/jira/connect/enterprise/app/${uuid}`,
+		data,
+		success: function (response) {
+			if (response.success) {
+				AP.history.back();
+			} else {
+				handleFormErrors(isUpdate);
+			}
+		}
+	});
 }
 
+$(document).ready(function() {
+	const update = document.getElementById("Update");
+	const isUpdate = update && update.innerText === "Update";
+
+	// Display the filename and make the data valid so users don't need to upload their pem file every time
+	if (isUpdate) {
+		$(".jiraManualAppCreation__formNoFileUploaded").hide();
+		$(".jiraManualAppCreation__formFileUploaded").css('display', 'flex');
+		$("#privateKeyFile").attr("data-aui-validation-state", "valid");
+	}
+});
+
 AJS.$("#jiraManualAppCreation__form").on("aui-valid-submit", (event) => {
-  event.preventDefault();
-  const form = event.target;
-  const csrf = $("#_csrf").val();
-  const data = $(form).serializeObject();
-  const isUpdate = $('input[type=submit]').val() === "Update";
-
-  // Reading the content of the file
-  const file = $("#privateKeyFile")[0].files[0];
-  const reader = new FileReader();
-  reader.readAsDataURL(file);
-
-  reader.onload = () => {
-    data.privateKey = reader.result;
-
-    AP.context.getToken((token) => {
-      data.jwt = token;
-      data._csrf = csrf;
-      data.jiraHost = jiraHost;
-
-      if (isUpdate) {
-        // TODO: Do a put request to update the existing app
-      } else {
-         $.post("/jira/connect/enterprise/app", data, (response, _status, result) => {
-          if (result.status === 201) {
-            // TODO: This doesn't work, will be done in ARC-1565
-            const child = openChildWindow(`/session/github/${response.data.uuid}/configuration?ghRedirect=to`);
-            child.window.jiraHost = jiraHost;
-            child.window.jwt = token;
-          }
-        });
-      }
-    });
-  };
+	event.preventDefault();
+	const form = event.target;
+	const data = $(form).serializeObject();
+	const updateElement = document.getElementById("Update");
+	const isUpdate = updateElement && updateElement.innerText === "Update";
+	const uuid = $(event.target).data("app-uuid");
+	const appName = $(event.target).data("app-appname");
+	const existingPrivateKey = $(event.target).data("app-privatekey");
```

Comment:
**IMPORTANT**

**NO NEED TO CHANGE IN THIS PR THOUGH**

I am concern from security point of view. I believe it is better to transmit the sensitive data as less as possible.

So in theory, we can omit sending the privateKey or webhook secret or whatever down to client browser for update. 

----

But I maybe misunderstood the code here. Let me know if I am wrong or need me to explain more.

Reply:
Yeah I had been thinking the same thing. You okay with me creating a follow-up ticket to look at an alternate solution for this?